# This notebook will contain the capstone project for IBM Data Science course

# Part 1: Import Toronto Neighborhoods from Wikipedia

Start by installing required libraries

In [1]:
#!pip install bs4
#!pip install requests

Import required libraries

In [2]:
from bs4 import BeautifulSoup
import requests
import pandas as pd
import numpy as np

Make the request to get the wikipedia page <br/>
Use beautiful soup to extract table with postal codes

In [3]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
data = requests.get(url).text
soup = BeautifulSoup(data,"html5lib")
tables = soup.find_all('table')

In [4]:
for i,table in enumerate (tables):
    if "Postal Code" in str(table.th):
        table_index=i
        break
print("The table index with postal codes is:", table_index)

The table index with postal codes is: 0


Put the table in a dataframe

In [5]:
ca_postalcodes = pd.read_html(str(tables[table_index]),flavor='bs4')[0]
ca_postalcodes.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

Remove the Borough Not assigned rows from the data frame

In [6]:
ca_postalcodes.drop(ca_postalcodes[ca_postalcodes['Borough']=="Not assigned"].index,inplace=True)
ca_postalcodes.reset_index(drop=True,inplace=True)
ca_postalcodes.head()

Postal Code           Borough                                Neighbourhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

#### Below is dummy code in case we need to rename "not assigned" Neighborhoods by Borough name, and group Neighborhoods with same postal code <br/>
PS: It seems the wiki table has been cleaned up since the creation of this assignment

In [7]:
bla={"postal code":["M1","M1","M2","M2"],"borough":["north york","north york","toronto","toronto"],"neigh":["park","not assigned","regent","not assigned"]}
df=pd.DataFrame(bla)
not_assigned=df.index[df['neigh']=='not assigned'].tolist()
df.loc[not_assigned,['neigh']]=df.loc[not_assigned,['borough']].values
df=df.groupby(['postal code','borough'])['neigh'].apply(','.join).reset_index()
df

postal code     borough            neigh
0          M1  north york  park,north york
1          M2     toronto   regent,toronto

In [8]:
ca_postalcodes.shape

(103, 3)

# Part 2: Import geolocation data to add to dataframe

In [9]:
!pip install geocoder
import geocoder

In [10]:
############### Having requests denied with geocoder.google, and not all results with geocoder.osm
#lat=[]
#lon=[]
#i=0
#while i<10:
#    for pc in zip(ca_postalcodes['Postal Code']):
#        print(pc)
#        # initialize your variable to None
#        lat_lng_coords = None
#        # loop until you get the coordinates
#        while(lat_lng_coords is None):
#            print(pc[0])
#            g = geocoder.osm('{}, Toronto, Ontario'.format(pc[0]))
#            lat_lng_coords = g.latlng
#        print(i)
#        lat.append(lat_lng_coords[0])
#        lon.append(lat_lng_coords[1])
#        i=i+1
#        print(i)
#    
#ca_postalcodes['Latitude']=lat
#ca_postalcodes['Longitude']=lon
#    
#ca_postalcodes.head()

We will get the coordinates from the csv file mentioned in the instructions

In [11]:
!wget -q -O 'Geospatial_data.csv' http://cocl.us/Geospatial_data
print('Data downloaded!') 

Data downloaded!


In [12]:
ca_lat_lon=pd.read_csv('Geospatial_data.csv')
ca_lat_lon.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

We apply a merge on key = "Postal Code" so pandas will the match Postal codes on both sides 

In [13]:
ca_merged=pd.merge(ca_postalcodes,ca_lat_lon,on="Postal Code")
ca_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

We still have the same number of Postal Codes (103) but now with 2 more columns: Latitude and Longitude

In [14]:
ca_merged.shape

(103, 5)

# Part 3 : Clustering neighborhoods in Toronto

We take only the boroughs with Toronto in the name

In [15]:
toronto_df=ca_merged[ca_merged['Borough'].str.contains('Toronto')]
toronto_df.reset_index(drop=True,inplace=True)
toronto_df.head()

Postal Code           Borough                                Neighbourhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.676357 -79.293031

In [16]:
toronto_df.shape

(40, 5)

In [17]:
address = 'Toronto, Ontario'

g = geocoder.osm(address)
latitude = g.latlng[0]
longitude = g.latlng[1]
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [18]:
!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

Solving environment: - 
The environment is inconsistent, please check the package plan carefully
The following packages are causing the inconsistency:

  - defaults/noarch::ibm-wsrt-py37main-main==custom=1902
  - conda-forge/linux-64::pytorch==1.8.0=cpu_py37hafa7651_0
  - defaults/noarch::ibm-wsrt-py37main-keep==0.0.0=1902
done

# All requested packages already installed.



Let's see the neighborhoods in question

In [19]:
# create map of Manhattan using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(toronto_df['Latitude'], toronto_df['Longitude'], toronto_df['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [20]:
CLIENT_ID = 'T5KYZG2JQ3MD3DTLKLPZTIH14NNGUFCZYG2SLXX00AX1YZMX' # your Foursquare ID
CLIENT_SECRET = '4ZEBMQCOLC23I1RXUBE0YG21V3L4WYLTHAEQUFXF10T3GFHL' # your Foursquare Secret
ACCESS_TOKEN = 'GVSKWL2XMRYNIGVM2ADI203H0I2FI5IHJWIOCH2GH0XUEZKB' # your FourSquare Access Token
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

Let's get the function from the NY lab to get nearby venues for all the neighborhoods

In [21]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [22]:
toronto_venues=getNearbyVenues(toronto_df['Neighbourhood'], toronto_df['Latitude'], toronto_df['Longitude'])
print(toronto_venues.shape)
toronto_venues.head()

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Runnymede, The Junction, Weston-Pellam Park, Carlton Village
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harb

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0  Regent Park, Harbourfront                43.65426               -79.360636   
1  Regent Park, Harbourfront                43.65426               -79.360636   
2  Regent Park, Harbourfront                43.65426               -79.360636   
3  Regent Park, Harbourfront                43.65426               -79.360636   
4  Regent Park, Harbourfront                43.65426               -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3     Body Blitz Spa East       43.654735       -79.359874   
4          Impact Kitchen       43.656369       -79.356980   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3                  Spa  
4           Restaurant

In [23]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Airport  Airport Food Court  Airport Gate  \
0  Regent Park, Harbourfront        0                   0             0   
1  Regent Park, Harbourfront        0                   0             0   
2  Regent Park, Harbourfront        0                   0             0   
3  Regent Park, Harbourfront        0                   0             0   
4  Regent Park, Harbourfront        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  ...  Theme Restaurant  Tibetan Restaurant  \
0             0         0  ...                 0                   0   
1             0         0  ...                 0                   0   
2             0         0  ...                 0                   0   
3             0         0  ...                 0                   0   
4             0         0  ...                 0                   0   

   Toy / Game Store  Trail  Train Station  Vegetarian / Vegan Restaurant  \
0                 0      0              0                              0   
1                 0      0              0                              0   
2                 0      0              0                              0   
3                 0      0              0                              0   
4                 0      0              0                              0   

   Video Game Store  Vietnamese Restaurant  Wine Bar  Yoga Studio  
0                 0                      0         0            0  
1                 0                      0         0            0  
2                 0                      0         0            0  
3                 0                      0         0            0  
4                 0                      0         0            0  

[5 rows x 229 columns]

In [24]:
# grouping by neighborhood and getting the mean of frequency of occurence of each category
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood   Airport  \
0                                         Berczy Park  0.000000   
1        Brockton, Parkdale Village, Exhibition Place  0.000000   
2   Business reply mail Processing Centre, South C...  0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...  0.066667   
4                                  Central Bay Street  0.000000   
5                                            Christie  0.000000   
6                                Church and Wellesley  0.000000   
7                      Commerce Court, Victoria Hotel  0.000000   
8                                          Davisville  0.000000   
9                                    Davisville North  0.000000   
10                       Dufferin, Dovercourt Village  0.000000   
11             First Canadian Place, Underground city  0.000000   
12    Forest Hill North & West, Forest Hill Road Park  0.000000   
13                           Garden District, Ryerson  0.000000   
14  Harbourfront East, Union Station, Toronto Islands  0.000000   
15                      High Park, The Junction South  0.000000   
16                     India Bazaar, The Beaches West  0.000000   
17          Kensington Market, Chinatown, Grange Park  0.000000   
18                                      Lawrence Park  0.000000   
19                           Little Portugal, Trinity  0.000000   
20                        Moore Park, Summerhill East  0.000000   
21                  North Toronto West, Lawrence Park  0.000000   
22                             Parkdale, Roncesvalles  0.000000   
23        Queen's Park, Ontario Provincial Government  0.000000   
24                          Regent Park, Harbourfront  0.000000   
25                           Richmond, Adelaide, King  0.000000   
26                                           Rosedale  0.000000   
27                                           Roselawn  0.000000   
28                                 Runnymede, Swansea  0.000000   
29  Runnymede, The Junction, Weston-Pellam Park, C...  0.000000   
30                                     St. James Town  0.000000   
31                        St. James Town, Cabbagetown  0.000000   
32                                     Stn A PO Boxes  0.000000   
33                                    Studio District  0.000000   
34  Summerhill West, Rathnelly, South Hill, Forest...  0.000000   
35                The Annex, North Midtown, Yorkville  0.000000   
36                                        The Beaches  0.000000   
37                       The Danforth West, Riverdale  0.000000   
38           Toronto Dominion Centre, Design Exchange  0.000000   
39                     University of Toronto, Harbord  0.000000   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0             0.000000      0.000000        0.000000              0.0   
1             0.000000      0.000000        0.000000              0.0   
2             0.000000      0.000000        0.000000              0.0   
3             0.066667      0.066667        0.066667              0.2   
4             0.000000      0.000000        0.000000              0.0   
5             0.000000      0.000000        0.000000              0.0   
6             0.000000      0.000000        0.000000              0.0   
7             0.000000      0.000000        0.000000              0.0   
8             0.000000      0.000000        0.000000              0.0   
9             0.000000      0.000000        0.000000              0.0   
10            0.000000      0.000000        0.000000              0.0   
11            0.000000      0.000000        0.000000              0.0   
12            0.000000      0.000000        0.000000              0.0   
13            0.000000      0.000000        0.000000              0.0   
14            0.000000      0.000000        0.000000              0.0   
15            0.000000      0.000000        0.000000              0.0   
16            0.000000      0.000000        0.00000

#### Time to do some clustering

In [25]:
#!pip install sklearn
#from sklearn.cluster import KMeans

In [26]:
from sklearn.cluster import KMeans

In [27]:
# set number of clusters
kclusters = 4

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=3).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

In [28]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [29]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...            Comic Shop   
3  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery          Cocktail Bar           Cheese Shop   
1                Bakery        Breakfast Spot           Coffee Shop   
2         Auto Workshop            Skate Park         Burrito Place   
3               Airport    Airport Food Court              Boutique   
4    Italian Restaurant        Sandwich Place                  Café   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Farmers Market            Restaurant    Seafood Restaurant   
1                   Gym         Burrito Place          Climbing Gym   
2  Fast Food Restaurant        Farmers Market    Light Rail Station   
3         Boat or Ferry      Sculpture Garden                   Bar   
4          Burger Joint   Japanese Restaurant           Salad Place   

  8th Most Common Venue   9th Most Common Venue 10th Most Common Venue  
0              Pharmacy                Beer Bar              Nightclub  
1     Convenience Store  Furniture / Home Store          Grocery Store  
2                Garden                 Brewery             Restaurant  
3   Rental Car Location             Coffee Shop           Airport Gate  
4       Bubble Tea Shop         Thai Restaurant                  Diner

In [30]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_df

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough                                Neighbourhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               0           Coffee Shop   
1  43.662301 -79.389494               0           Coffee Shop   
2  43.657162 -79.378937               0        Clothing Store   
3  43.651494 -79.375418               0                  Café   
4  43.676357 -79.293031               0          Neighborhood   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery                  Park        Breakfast Spot   
1                   Gym      Sushi Restaurant                 Diner   
2           Coffee Shop                 Hotel    Italian Restaurant   
3           Coffee Shop             Gastropub          Cocktail Bar   
4     Health Food Store                 Trail                   Pub   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                   Pub                  Café            Restaurant   
1   Fried Chicken Joint         Burrito Place                  Café   
2   Japanese Restaurant                  Café        Cosmetics Shop   
3                   Gym    Italian Restaurant                 Hotel   
4               Airport    Miscellaneous Shop   Monument / Landmark   

             8th Most Common Venue       9th Most Common Venue  \
0                          Theater                     Brewery   
1               College Auditorium                    Creperie   
2                  Bubble Tea Shop   Middle Eastern Restaurant   
3                   Lingerie Store         Moroccan Restaurant   
4  Molecular Gastronomy Restaurant  Modern European Restaurant   

  10th Most Common Venue  
0                    Spa  
1    Distribution Center  
2              Bookstore  
3         Farmers Market  
4     Mexican Restaurant

In [31]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

In [32]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters